### PR TITLE
Add Data and Coding Blob variants for data plane

### PR DESCRIPTION
**Pull request by mark-solana**
#### Problem
Part of [Issue 3981](https://github.com/solana-labs/solana/issues/3981) 

#### Summary of Changes
Create a new module `solana::blob`. All of it is re-exported by `solana::packet` so code can migrate more easily.

Add `BlobHeader` enum and implement manual serialization/deserialization
from slices

Add `CodingHeader` and `DataHeader` structs.

Add `CodingBlob` and `DataBlob` structs with the deserialized headers.
`DataBlob`s do not expose the last `BLOB_HEADER_SIZE` bytes of the buffer since erasure needs that space to remain empty.

Add methods to convert between `Blob` <=> `CodingBlob` and `Blob` <=> `DataBlob`
Fixes #

cc @rob-solana @garious Does this look like the right direction to move in? Not sure if the types are useful as is.

### CODE REVIEW

**Comment by garious** [comment]
No reviewers on draft PRs please. Mention us instead. https://github.com/solana-labs/solana/blob/master/CONTRIBUTING.md#draft-pull-requests

**Comment by garious** [comment]
> Part of Issue 3981

Which part?

**Comment by mark-solana** [comment]
Sorry about that. Updated the PR.

Adding the mentioned types and methods. I haven't migrated any client code yet as I'm unsure about the general approach.

Also I decided against implementing `Serialize` and `Deserialize` for any of the types involved since none of them have much use for actually being serialized for storage or transport.

**Comment by garious** [comment]
Why the manual serialization/deserialization? Performance?

**Comment by mark-solana** [comment]
It can't be automatically derived and it's less code to do it manually than implementing the traits from serde.

**Comment by garious** [comment]
> It can't be automatically derived

Why's that?

**Comment by garious** [comment]
Why not make `Blob` an enum or a trait? enum if using a closed type isn't a concern. trait if an open type is required.

**Comment by garious** [comment]
@mark-solana, any news here?

File: core/src/blob.rs [hunk 1]
```diff
@@ -0,0 +1,921 @@
...
+use std::fmt;
+use std::io;
+use std::io::Cursor;
+use std::io::Write;
+use std::mem::size_of;
+use std::net::{SocketAddr, UdpSocket};
+use std::ops::{Deref, DerefMut};
+use std::sync::{Arc, RwLock};
+
+pub type SharedBlob = Arc<RwLock<Blob>>;
```
Comment (rob-solana) [inline core/src/blob.rs]: are we still using this type aliasing style for new code?

**Comment by mark-solana** [comment]
closing for now

**Pull request closed without merging**